### PR TITLE
Enhance json parser error logging to better track Istio Proxy error message

### PR DESCRIPTION
**Pull request by kaisun2000**
### Description

Currently the inter Druid communication via rest endpoints is based on json formatted payload. Upon parsing error, there is only a generic exception stating expected json token type and current json token type. There is no detailed error log about the content of the payload causing the violation. 

In the micro-service world, the trend is to deploy the Druid servers in k8 with the mesh network. Often the istio proxy or other proxies is used to intercept the network connection between Druid servers. The proxy may give error messages for various reasons. These error messages are not expected by the json parser. The generic error message from Druid can be very misleading as the user may think the message is based on the response from the other Druid server. 

For example, this is an example of mysterious error message

```
QueryInterruptedException{msg=Next token wasn't a START_ARRAY, was[VALUE_STRING] from url[http://xxxxx:8088/druid/v2/], code=Unknown exception, class=org.apache.druid.java.util.common.IAE, host=xxxxx:8088}"
```

While the context of the message is the following from the proxy when it can't tunnel the network connection.

```
pstream connect error or disconnect/reset before header
```

So this very simple PR is just to enhance the logging and get the real underlying message printed out. This would save a lot of head scratching time if Druid is deployed with mesh network. 

#### Release note

<hr>
Improved: enhanced the json parser unexpected token logging with the context of the expected VALUE_STRING token. This way, it is easier to track to mesh/proxy network error messages, instead misleading user to track the Druid server rest endpoint responses.
<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/client/JsonParserIterator.java [hunk 1]
```diff
@@ -185,6 +185,12 @@ private void init()
...
           throw convertException(jp.getCodec().readValue(jp, QueryException.class));
         } else {
+          if (nextToken == JsonToken.VALUE_STRING) {
+            LOG.error(
+                "Next Token value is of type VALUE_STRING with content [%s], not as expected START_ARRAY",
+                jp.getText()
+            );
+          }
           throw convertException(
               new IAE("Next token wasn't a START_ARRAY, was[%s] from url[%s]", jp.getCurrentToken(), url)
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/client/JsonParserIterator.java:195]: Should we add `jp.getText()` in the exception. That way its a much cleaner implementation?
We should chom jp.getText() to around 200 chars before adding it to the message though.

**Review by cryptoe — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/druid/client/JsonParserIterator.java (hunk 1 above)
Comment (kaisun2000) [reply server/src/main/java/org/apache/druid/client/JsonParserIterator.java]: added a unit test and chop the error message at most to 192.

File: server/src/main/java/org/apache/druid/client/JsonParserIterator.java [hunk 2]
```diff
@@ -185,8 +185,15 @@ private void init()
         } else if (nextToken == JsonToken.START_OBJECT) {
           throw convertException(jp.getCodec().readValue(jp, QueryException.class));
         } else {
+          String errMsg = jp.getValueAsString();
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/client/JsonParserIterator.java:188]: Since jp.getValueAsString() can return null 
```
Method that will try to convert value of current token to a String. JSON Strings map naturally; scalar values get converted to their textual representation. If representation can not be converted to a String value (including structured types like Objects and Arrays and null token), default value of null will be returned; no exceptions are thrown.
Since:
2.1
```
lets add a null check here.

**Review by cryptoe — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/druid/client/JsonParserIterator.java (hunk 2 above)
Comment (kaisun2000) [reply server/src/main/java/org/apache/druid/client/JsonParserIterator.java:188]: good catch. added checking null string and upon null value, not doing the string limitting to 192 length check.

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java [hunk 3]
```diff
@@ -316,8 +316,41 @@ private Query<?> mockQuery(String queryId, long timeoutAt)
     }
   }
 
+  public static class IAEExceptionConversionTest
+  {
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private String meshErrMsg = "pstream connect error or disconnect/reset before header";
```
Comment (cryptoe) [inline server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java:324]: Also, can we add a test case with null value. 
```suggestion
    private String errorMessage = "pstream connect error or disconnect/reset before header";
```

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java [hunk 4]
```diff
@@ -316,8 +316,41 @@ private Query<?> mockQuery(String queryId, long timeoutAt)
...
+
+    @Test
+    public void testMeshProxyError()
```
Comment (cryptoe) [inline server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java:327]:
```suggestion
    public void testParsingError()
```

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java [hunk 5]
```diff
@@ -316,8 +316,41 @@ private Query<?> mockQuery(String queryId, long timeoutAt)
...
+      iterator.hasNext();
+    }
+  }
+
   private static InputStream mockErrorResponse(Exception e) throws JsonProcessingException
   {
     return new ByteArrayInputStream(OBJECT_MAPPER.writeValueAsBytes(e));
   }
+
+  private static InputStream mockMeshProxyResponse(String errMsg) throws JsonProcessingException
```
Comment (cryptoe) [inline server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java:352]:
```suggestion
  private static InputStream mockErrorResponse(String errMsg) throws JsonProcessingException
```

**Review by cryptoe — COMMENTED** [review]
After the changes its a +1 from my side.

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java (hunk 5 above)
Comment (kaisun2000) [reply server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java]: changed as suggested

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java (hunk 4 above)
Comment (kaisun2000) [reply server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java]: changed as suggested.

File: server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java (hunk 3 above)
Comment (kaisun2000) [reply server/src/test/java/org/apache/druid/client/JsonParserIteratorTest.java]: added the null test case as suggested.

**Review by cryptoe — APPROVED** [review]
Thanks @kaisun2000 for the contribution. 
Change LGTM.
